README.md
> Starting with [`bpmn-js@11.2`](https://github.com/bpmn-io/bpmn-js/blob/develop/CHANGELOG.md#1120) this is now integrated into the modeling core.

# bpmn-js-connectors-extension

[![CI](https://github.com/bpmn-io/bpmn-js-connectors-extension/actions/workflows/CI.yml/badge.svg)](https://github.com/bpmn-io/bpmn-js-connectors-extension/actions/workflows/CI.yml)

This module extends [bpmn-js](https://github.com/bpmn-io/bpmn-js) with an element templates everywhere modeling experience. 

[![screenshot](./resources/screenshot.png)](https://potential-winner-9f6a854d.pages.github.io/)


## Features

Built on top of [element templates](https://docs.camunda.io/docs/components/modeler/camunda-modeler/element-templates/camunda-modeler-element-templates/):

* Append template tasks during modeling
* Upgrade template task via replace menu
* (Optionally) Upgrade template tasks via properties panel
* Modern replace and append with search + keyboard support
* Documentation links and icons
* Visually distinguish template tasks from standard BPMN tasks on the canvas

## Resources

* [Issues](https://github.com/bpmn-io/bpmn-js-connectors-extension/issues)
* [Changelog](./CHANGELOG.md)


## Use Extension

Install via npm:

```
npm install bpmn-js-connectors-extension
```

Use in your [bpmn-js powered editor](https://github.com/bpmn-io/bpmn-js):

```javascript
import ConnectorsExtensionModule from 'bpmn-js-connectors-extension';

import 'bpmn-js-connectors-extension/dist/connectors-extension.css';

const modeler = new BpmnModeler({
  additionalModules: [
    ...,
    ConnectorsExtensionModule,
  ],
  moddleExtensions: {
    zeebe: ZeebeModdle
  }
});
```

The extension relies on `elementTemplates` to be provided via additional setup, i.e. [camunda-bpmn-js](https://github.com/camunda/camunda-bpmn-js).

See [`example`](./example) for a full featured demo or [check it out in action](https://bpmn-io.github.io/bpmn-js-connectors-extension/).


## Configuration

Customize this extension via the `connectorsExtension` config: 

| Option | Impact |
| :--- | :--- |
| `appendAnything=false` | If `true`, append and create anything modeling experience |
| `elementTemplateChooser=true` | If `false`, do not integrate with the properties panel element template chooser |


## Related

* [`@bpmn-io/element-template-icon-renderer`](https://github.com/bpmn-io/element-template-icon-renderer) - A renderer for task icons
* [`@bpmn-io/element-template-chooser`](https://github.com/bpmn-io/element-template-chooser) - A simple element template chooser that plugs into the [BPMN properties panel](https://github.com/bpmn-io/bpmn-js-properties-panel)


## Run locally

To run the demo application, execute:

```
npm start
```


CHANGELOG.md
# Changelog

All notable changes to [bpmn-js-connectors-extension](https://github.com/bpmn-io/bpmn-js-connectors-extension) are documented here. We use [semantic versioning](http://semver.org/) for releases.

## Unreleased

___Note:__ Yet to be released changes appear here._

## 0.4.6

* `DOCS`: various documentation updates

## 0.4.5

* `FIX`: key replace menu entries by ID ([#55](https://github.com/bpmn-io/bpmn-js-connectors-extension/pull/55))

## 0.4.4

* `FIX`: revert [`6189a49`](https://github.com/bpmn-io/bpmn-js-connectors-extension/commit/6189a499522250582584475bfc82ea6a0c544c50) due to [#54](https://github.com/bpmn-io/bpmn-js-connectors-extension/issues/54)

## 0.4.3

* `FIX`: force create mode when appending boundary event ([#52](https://github.com/bpmn-io/bpmn-js-connectors-extension/pull/52))
* `DEPS`: bump modeling dependencies

## 0.4.2

* `DEPS`: bump modeling dependencies

## 0.4.1

* `FIX`: always upper case `Connector`

## 0.4.0

* `FEAT`: only offer latest template versions for selection ([#38](https://github.com/bpmn-io/bpmn-js-connectors-extension/issues/38))
* `FIX`: gracefully handle missing template `category` ([#43](https://github.com/bpmn-io/bpmn-js-connectors-extension/issues/43))
* `FIX`: prevent appending templates to connections ([#46](https://github.com/bpmn-io/bpmn-js-connectors-extension/issues/46))
* `FIX`: do not trigger append on external labels ([#39](https://github.com/bpmn-io/bpmn-js-connectors-extension/issues/39))
* `FIX`: complete direct editing when opening popup menu ([#40](https://github.com/bpmn-io/bpmn-js-connectors-extension/issues/40))

## 0.3.1

* `FIX`: handle missing target in replace options ([`cba23f08`](https://github.com/bpmn-io/bpmn-js-connectors-extension/commit/cba23f084cfca854436da305dcf647dc782c94c7))

## 0.3.0

* `FEAT`: insert unlink entry late ([#35](https://github.com/bpmn-io/bpmn-js-connectors-extension/pull/35))

## 0.2.0

* `FEAT`: allow unlink of template via replace menu ([#32](https://github.com/bpmn-io/bpmn-js-connectors-extension/issues/32))

## 0.1.1

* `FIX`: make entry ids for menu actions unique

## 0.1.0

* `FEAT`: require official [template icon renderer](https://github.com/bpmn-io/element-templates-icons-renderer)

### Breaking Change

* Rendering of template icons on the canvas is now a concern outside this library.

## 0.0.13

* `FIX`: correctly handle missing `appendAnything` feature toggle

## 0.0.12

* `FEAT`: support template icons ([#26](https://github.com/bpmn-io/bpmn-js-connectors-extension/pull/26))
* `FEAT`: support categories and search ([#18](https://github.com/bpmn-io/bpmn-js-connectors-extension/pull/18))
* `FEAT`: improve mixed mouse + keyboard selection ([#19](https://github.com/bpmn-io/bpmn-js-connectors-extension/pull/19))
* `FEAT`: improve existing tools integration ([#17](https://github.com/bpmn-io/bpmn-js-connectors-extension/pull/17))
* `FEAT`: visually hide search if less than five elements ([#16](https://github.com/bpmn-io/bpmn-js-connectors-extension/pull/16))
* `FEAT`: optionally support create/append anything ([#14](https://github.com/bpmn-io/bpmn-js-connectors-extension/pull/14))
* `FEAT`: add append (`a`), replace (`r`) and create (`n`) keyboard bindings ([#15](https://github.com/bpmn-io/bpmn-js-connectors-extension/pull/15))
* `FIX`: correct muted entry sizing ([#24](https://github.com/bpmn-io/bpmn-js-connectors-extension/pull/24))
* `FIX`: correct scroll into on Firefox ([#23](https://github.com/bpmn-io/bpmn-js-connectors-extension/pull/23))
* `FIX`: correct pass through of `documentationRef` ([#28](https://github.com/bpmn-io/bpmn-js-connectors-extension/pull/28))
* `FIX`: prevent click-through on documentation ref ([#28](https://github.com/bpmn-io/bpmn-js-connectors-extension/pull/28))
* `FIX`: prevent ENTER propagation on append

## 0.0.11

* `CHORE`: remove debug logging ([#13](https://github.com/bpmn-io/bpmn-js-connectors-extension/pull/13))

## 0.0.10

* `FEAT`: integrate with other editing components ([#10](https://github.com/bpmn-io/bpmn-js-connectors-extension/pull/10))
* `FEAT`: support create anywhere from append menu ([#11](https://github.com/bpmn-io/bpmn-js-connectors-extension/pull/11))

## 0.0.9

* `FEAT`: move keyboard selected entries into viewport ([`87fd87fe`](https://github.com/bpmn-io/bpmn-js-connectors-extension/commit/87fd87feca03a4d0e40c1e69c2c7001df67ced98))
* `FEAT`: show docs link based on `documentationRef` ([`1fb704bb`](https://github.com/bpmn-io/bpmn-js-connectors-extension/commit/1fb704bb55e345114e6dee68f3bab48ccc9e632b))
* `CHORE`: position element template chooser based on available height ([`ec5c9055`](https://github.com/bpmn-io/bpmn-js-connectors-extension/commit/ec5c90550418b043faf3c2a257d4e2872d499884))

## ...

Check `git log` for earlier history.


src/ConnectorsExtension.js
const appendPath = `
<path d="M1.3 3.4c.3 0 .5-.2.5-.5s-.2-.4-.5-.4c-.2 0-.4.1-.4.4 0 .3.2.5.4.5zM3 3.4c.2 0 .4-.2.4-.5s-.2-.4-.4-.4c-.3 0-.5.1-.5.4 0 .3.2.5.5.5zM4.6 3.4c.2 0 .4-.2.4-.5s-.2-.4-.4-.4c-.3 0-.5.1-.5.4 0 .3.2.5.5.5z"/>
`;

const appendConnectorSvg = `
<svg width="22" height="22" viewBox="0 0 22 22" fill="none" xmlns="http://www.w3.org/2000/svg">
<path fill-rule="evenodd" clip-rule="evenodd" d="M15.8125 11.0069C14.7767 11.0079 13.7689 11.343 12.9388 11.9625L10.0375 9.06127C10.6643 8.2372 11.0026 7.22976 11 6.19439C11.0056 3.96215 9.47542 2.01909 7.30408 1.50118C5.13274 0.983268 2.89029 2.02648 1.88772 4.02092C0.885144 6.01536 1.38567 8.43742 3.09667 9.87109C4.80766 11.3048 7.27994 11.3737 9.06813 10.0375L11.9625 12.9388C10.6267 14.7264 10.6952 17.1979 12.1278 18.9089C13.5605 20.6199 15.9814 21.1216 17.976 20.1207C19.9706 19.1199 21.0155 16.8791 20.5 14.7079C19.9846 12.5366 18.0441 11.0045 15.8125 11.0069ZM2.75001 6.19439C2.75001 4.29591 4.28903 2.75689 6.18751 2.75689C8.08599 2.75689 9.62501 4.29591 9.62501 6.19439C9.62501 8.09287 8.08599 9.63189 6.18751 9.63189C4.28903 9.63189 2.75001 8.09287 2.75001 6.19439Z" fill="#22242A"/>
</svg>
`;

const appendConnectorImageUrl = 'data:image/svg+xml;utf8,' + encodeURIComponent(appendConnectorSvg);

const appendSvg = `
<svg width="22" height="22" viewBox="0 0 5.82 5.82" xmlns="http://www.w3.org/2000/svg">
  ${appendPath}
</svg>
`;

const appendImageUrl = 'data:image/svg+xml;utf8,' + encodeURIComponent(appendSvg);

const createSvg = `
<svg width="46" height="46" viewBox="-2 -2 9.82 9.82" xmlns="http://www.w3.org/2000/svg">
  ${appendPath}
</svg>
`;

const createImageUrl = 'data:image/svg+xml;utf8,' + encodeURIComponent(createSvg);

// workaround for
// https://github.com/camunda/camunda-bpmn-js/issues/87
const LOWER_PRIORITY = 499;


export default function ConnectorsExtension(
    config,
    injector, create,
    contextPad, translate,
    elementTemplatesLoader,
    replaceMenu, appendMenu,
    createMenu, palette) {

  this._create = create;
  this._contextPad = contextPad;
  this._translate = translate;
  this._elementTemplatesLoader = elementTemplatesLoader;
  this._replaceMenu = replaceMenu;
  this._appendMenu = appendMenu;
  this._createMenu = createMenu;

  this._autoPlace = injector.get('autoPlace', false);

  this._config = config;

  contextPad.registerProvider(LOWER_PRIORITY, this);

  if (this._isAppendAnything()) {
    palette.registerProvider(LOWER_PRIORITY, this);
  }
}

ConnectorsExtension.$inject = [
  'config.connectorsExtension',
  'injector', 'create',
  'contextPad', 'translate',
  'elementTemplatesLoader',
  'replaceMenu', 'appendMenu',
  'createMenu', 'palette'
];

ConnectorsExtension.prototype.loadTemplates = function(templates) {
  this._elementTemplatesLoader.setTemplates(templates);
};

ConnectorsExtension.prototype._isAppendAnything = function() {
  return this._config && this._config.appendAnything || false;
};

ConnectorsExtension.prototype._getReplaceMenuPosition = function(element) {

  var Y_OFFSET = 5;

  var pad = this._contextPad.getPad(element).html;

  var padRect = pad.getBoundingClientRect();

  var pos = {
    x: padRect.left,
    y: padRect.bottom + Y_OFFSET
  };

  return pos;
};

ConnectorsExtension.prototype._getAppendMenuPosition = function(element) {

  var X_OFFSET = 5;

  var pad = this._contextPad.getPad(element).html;

  var padRect = pad.getBoundingClientRect();

  var pos = {
    x: padRect.right + X_OFFSET,
    y: padRect.top
  };

  return pos;
};

ConnectorsExtension.prototype.appendAnything = function(event, element) {

  const appendMenu = this._appendMenu;

  if (appendMenu.isEmpty(element)) {
    return;
  }

  const position = {
    ...(this._getAppendMenuPosition(element)),
    cursor: event && { x: event.x, y: event.y }
  };

  return appendMenu.open(element, position).then(result => {

    if (!result) {
      return;
    }

    const {
      event,
      newElement,
      dragstart = false
    } = result;

    const createStart = (event, source, newElement) => {
      this._create.start(event, newElement, {
        source
      });
    };

    const append = !dragstart && this._autoPlace
      ? (event, source, newElement) => {
        this._autoPlace.append(source, newElement);
      }
      : createStart;

    append(event, element, newElement);
  }).catch(error => {
    if (error !== 'user-canceled') {
      console.error('append-anything :: error', error);
    }
  });
};

ConnectorsExtension.prototype.replaceAnything = function(event, element) {

  const replaceMenu = this._replaceMenu;

  if (replaceMenu.isEmpty(element)) {
    return;
  }

  const position = {
    ...(this._getReplaceMenuPosition(element)),
    cursor: event && { x: event.x, y: event.y }
  };

  return replaceMenu.open(element, position).catch(error => {
    if (error !== 'user-canceled') {
      console.error('replace-anything :: error', error);
    }
  });
};

ConnectorsExtension.prototype.createAnything = function(event, position) {

  return this._createMenu.open(event, position).then(result => {

    const {
      event: _event,
      newElement
    } = result;

    this._create.start(
      _event instanceof MouseEvent ? _event : event,
      newElement
    );
  }).catch(error => {
    if (error !== 'user-canceled') {
      console.error('create-anything :: error', error);
    }
  });
};

ConnectorsExtension.prototype.getContextPadEntries = function(element) {

  const replaceMenu = this._replaceMenu;
  const translate = this._translate;

  if (element.labelTarget) {
    return null;
  }

  return (entries) => {

    // hook up new replace menu

    // clear replace menu
    if (replaceMenu.isEmpty(element)) {
      const {
        replace,
        ...restEntries
      } = entries;

      entries = restEntries;
    } else {

      entries = { ...entries };

      entries['replace'] = {
        group: 'edit',
        className: 'bpmn-icon-screw-wrench',
        title: translate('Change type'),
        action: {
          click: (event, element) => {
            this.replaceAnything(event, element);
          }
        }
      };
    }

    // only allow when appending task is allowed, too
    if (entries['append.append-task']) {

      entries = { ...entries };

      const appendAnything = this._isAppendAnything();

      if (appendAnything) {
        entries['append-anything'] = {
          group: 'model',
          imageUrl: appendImageUrl,
          title: translate('Append anything'),
          action: {
            click: (event, element) => {
              this.appendAnything(event, element);
            }
          }
        };
      } else {
        entries['append-connector'] = {
          group: 'model',
          imageUrl: appendConnectorImageUrl,
          title: translate('Append Connector'),
          action: {
            click: (event, element) => {
              this.appendAnything(event, element);
            }
          }
        };
      }
    }

    return entries;
  };
};

ConnectorsExtension.prototype.getPaletteEntries = function() {

  return {
    'create-anything': {
      group: 'anything',
      imageUrl: createImageUrl,
      title: this._translate('Create anything'),
      action: {
        click: (event) => {

          const cursor = event && { x: event.x, y: event.y };

          const position = event && {
            x: cursor.x + 35,
            y: cursor.y + 10,
            cursor
          };

          this.createAnything(event, position);
        }
      }
    }
  };
};